File: test_3_4/threading.py
import threading
import time
import sqlite3
from concurrent.futures import ThreadPoolExecutor

class ThreadMaker(threading.Thread):
    def __init__(self, id):
        super().__init__()
        self.id = id

    def run(self):
        for i in range(5):
            self.execute_query(i)
            time.sleep(2)

    def execute_query(self, param):
        for _ in range(10):
            conn = sqlite3.connect(":memory:")
            cursor = conn.cursor()
            cursor.execute("CREATE TABLE IF NOT EXISTS test (id INTEGER PRIMARY KEY, val TEXT)")
            cursor.execute("INSERT INTO test (val) VALUES ('example')")
            cursor.execute("SELECT * FROM test WHERE id = ?", (param,))
            cursor.fetchall()
            cursor.close()
            conn.close()

class Worker:
    def __init__(self, id):
        self.id = id

    def perform(self):
        for i in range(5):
            self.run_query(i)
            time.sleep(1.5)

    def run_query(self, parameter):
        conn = sqlite3.connect(":memory:")
        cursor = conn.cursor()
        cursor.execute("CREATE TABLE IF NOT EXISTS test (id INTEGER PRIMARY KEY, val TEXT)")
        cursor.execute("INSERT INTO test (val) VALUES ('sample')")
        cursor.execute("SELECT val FROM test WHERE id = ?", (parameter,))
        cursor.fetchall()
        cursor.close()
        conn.close()

def main():
    threads = [ThreadMaker(i) for i in range(5)]
    for t in threads:
        t.start()

    with ThreadPoolExecutor(max_workers=3) as executor:
        workers = [Worker(i) for i in range(5)]
        for w in workers:
            executor.submit(w.perform)

    for t in threads:
        t.join()

if __name__ == "__main__":
    main()

File: test_3_4/threading.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

class ThreadCreator extends Thread {
    private int id;

    public ThreadCreator(int id) {
        this.id = id;
    }

    public void run() {
        try {
            for (int i = 0; i < 5; i++) {
                performQuery(i);
                Thread.sleep(2000);
            }
        } catch (Exception e) {
        }
    }

    private void performQuery(int index) {
        for (int j = 0; j < 10; j++) {
            try {
                Connection conn = DriverManager.getConnection("jdbc:h2:mem:test");
                PreparedStatement ps = conn.prepareStatement("SELECT * FROM test_table WHERE id = ?");
                ps.setInt(1, j);
                ps.executeQuery();
                ps.close();
                conn.close();
            } catch (Exception e) {
            }
        }
    }
}

class TaskWorker implements Runnable {
    private int id;

    public TaskWorker(int id) {
        this.id = id;
    }

    public void run() {
        try {
            for (int i = 0; i < 5; i++) {
                queryDatabase(i);
                Thread.sleep(1500);
            }
        } catch (Exception e) {
        }
    }

    private void queryDatabase(int param) {
        try {
            Connection conn = DriverManager.getConnection("jdbc:h2:mem:test");
            PreparedStatement ps = conn.prepareStatement("SELECT name FROM test_table WHERE key = ?");
            ps.setInt(1, param);
            ps.executeQuery();
            ps.close();
            conn.close();
        } catch (Exception e) {
        }
    }
}

public class Application {
    public static void main(String[] args) throws InterruptedException {
        for (int i = 0; i < 5; i++) {
            new ThreadCreator(i).start();
        }

        ExecutorService pool = Executors.newFixedThreadPool(3);

        for (int i = 0; i < 5; i++) {
            pool.submit(new TaskWorker(i));
        }

        pool.shutdown();
        pool.awaitTermination(10, TimeUnit.MINUTES);
    }
}
